dialogue-core/src/main/java/com/palantir/dialogue/core/StickySessioning.java

```diff
@@ -0,0 +1,120 @@
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.palantir.dialogue.Channel;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import com.palantir.dialogue.TagKey;
+import java.io.InputStream;
+import java.util.Optional;
+
+public final class StickySessioning {
+    private static final TagKey<LimitedChannel> EXECUTED_ON = new TagKey<>(LimitedChannel.class);
+
+    private StickySessioning() {}
+
+    static LimitedChannel wrapWithRecordedChoiceSelection(LimitedChannel delegate) {
+        return new RecordedChoiceChannel(delegate);
+    }
+
+    static ListenableFuture<Response> addExecutedOnTag(
+            ListenableFuture<Response> responseFuture,
+            LimitedChannel channel) {
+        return Futures.transform(responseFuture,
+                response -> new Response() {
+                    @Override
+                    public InputStream body() {
+                        return response.body();
+                    }
+
+                    @Override
+                    public int code() {
+                        return response.code();
+                    }
+
+                    @Override
+                    public ListMultimap<String, String> headers() {
+                        return response.headers();
+                    }
+
+                    @Override
+                    public void close() {
+                        response.close();
+                    }
+
+                    @Override
+                    public <T> Optional<T> getTag(TagKey<T> tagKey) {
+                        if (tagKey == EXECUTED_ON) {
+                            T ret = (T) channel;
+                            return Optional.of(ret);
+                        }
+                        return response.getTag(tagKey);
+                    }
+                }, MoreExecutors.directExecutor());
+    }
+
+    private static final class RecordedChoiceChannel implements LimitedChannel {
+        private final LimitedChannel delegate;
+
+        private RecordedChoiceChannel(LimitedChannel delegate) {
+            this.delegate = delegate;
+        }
+
+        @Override
+        public Optional<ListenableFuture<Response>> maybeExecute(
+                Endpoint endpoint, Request request) {
+            return request.getTag(EXECUTED_ON).orElse(delegate).maybeExecute(endpoint, request);
+        }
+    }
+
+    private static final class StickySessionChannel implements Channel {
+        private final Channel delegate;
+
+        private FluentFuture<LimitedChannel> stuckChannel;
+
+        private StickySessionChannel(Channel delegate) {
+            this.delegate = delegate;
+        }
+
+        @Override
+        public synchronized ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
```

Comment:
If this is constructed in the way I described, this synchronized is not a problem.

Reply:
Would an explicit mapping in the response for "this is the Channel I executed on" be a bit less open?